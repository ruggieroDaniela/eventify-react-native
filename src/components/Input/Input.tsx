import { View, Text, StyleSheet, TouchableOpacity, TextInput } from "react-native";
import { theme } from "../../../utils/theme";
import { MaterialIcons } from "@expo/vector-icons";

export enum InputVariant {
  DEFAULT = "default",
  ARROW = "arrow",
}

interface InputProps {
  label: string;
  placeholder?: string;
  multiline?: boolean; 
  variant?: InputVariant;
  required?: boolean; 
  onPress?: () => void;
  value?: string, 
  onChangeValue?: (data: string) => void;
  testID?: string;
}

export function Input({
  label,
  placeholder,
  variant = InputVariant.DEFAULT,
  multiline = true, 
  required = true, 
  onPress,
  value, 
  onChangeValue,
  testID
}: InputProps) {
  return (
    <TouchableOpacity testID={testID} onPress={onPress} style={styles.container}>
      <Text style={[styles.label, { flex: placeholder ? 0.45 : 1 }]}>
        {label}{" "}
        {required && <Text style={styles.required}>*</Text>}
      </Text>
      
      {variant === InputVariant.DEFAULT 
       ? (<TextInput 
            multiline={multiline}
            numberOfLines={multiline ? 3 : 1}
            style={styles.placeholder}
            placeholder={placeholder}
            onChange={onPress}
            value={value}
            onChangeText={onChangeValue}
            testID={testID ? `${testID}-input` : undefined}
       />)
       : (placeholder && <Text style={styles.placeholder}>{placeholder}</Text>)
      }
      
      {variant === InputVariant.ARROW ? (
        <MaterialIcons name="keyboard-arrow-right" size={24} color="gray" />
      ) : null}
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    padding: 10,
    borderBottomWidth: 2,
    borderBottomColor: theme.colors["gray"],
    justifyContent: "space-between",
  },

  label: {
    fontWeight: "bold",
    fontFamily: "SF-Pro-Rounded-Bold",
  },

  placeholder: {
    flex: 1,
    color: "gray",
    fontFamily: "SF-Pro-Text-Regular",
  },
  required:{
    color: theme.colors['red']
  }
});
