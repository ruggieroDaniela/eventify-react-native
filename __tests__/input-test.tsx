import { fireEvent, render, screen } from '@testing-library/react-native';

import { Input, InputVariant } from '../src/components/Input/Input';

describe('Input Component', () => {
  test('renders input with label', () => {
    render(<Input label="Test Label" />);
    const labelElement = screen.getByText(/Test Label/);
    expect(labelElement).toBeTruthy();
  });
  
  test('displays arrow icon when variant is ARROW', () => {
    render(<Input label="Test Arrow" variant={InputVariant.ARROW} />);
    const labelElement = screen.getByText(/Test Arrow/);
    expect(labelElement).toBeTruthy();
    
    const iconText = screen.getByText('keyboard-arrow-right');
    expect(iconText).toBeTruthy();
  });
  
  test('calls onPress when the input container is clicked', () => {
    const mockOnPress = jest.fn();
  
    render(
      <Input 
        label="Click Test" 
        onPress={mockOnPress}
        testID="input-container" 
      />
    ); 
  
    // Usar el testID para encontrar el componente
    const touchable = screen.getByTestId('input-container');
    fireEvent.press(touchable);
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });
  
  test('calls onChangeValue when the input text changes', () => {
    const mockOnChangeValue = jest.fn(); 
  
    render(
      <Input
        label="Test"
        placeholder="Enter text"
        onChangeValue={mockOnChangeValue}
        value=""
        testID="change-test"
      />
    ); 
  
    const inputElement = screen.getByPlaceholderText('Enter text'); 
    fireEvent.changeText(inputElement, 'new text');
  
    expect(mockOnChangeValue).toHaveBeenCalledTimes(1);
    expect(mockOnChangeValue).toHaveBeenCalledWith('new text');
  });


  test('Custom input renders correctly', () => {
    const tree = render(<Input label="Custom Input" />);
    expect(tree).toMatchSnapshot();
  });
});