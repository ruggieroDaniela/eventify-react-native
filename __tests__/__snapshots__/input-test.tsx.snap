// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Custom input renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderBottomColor": "#E9E8E8",
      "borderBottomWidth": 2,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "padding": 10,
    }
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "SF-Pro-Rounded-Bold",
          "fontWeight": "bold",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    Custom Input
     
    <Text
      style={
        {
          "color": "#FF0000",
        }
      }
    >
      *
    </Text>
  </Text>
  <TextInput
    multiline={true}
    numberOfLines={3}
    style={
      {
        "color": "gray",
        "flex": 1,
        "fontFamily": "SF-Pro-Text-Regular",
      }
    }
  />
</View>
`;
